File: server_pythonver/tool/web_search_service.py
import os
import datetime
from urllib.parse import urlparse

# LangChainのインポート (インストールが必要)
# from langchain_community.tools import TavilySearchResults
# from langchain_community.tools import DuckDuckGoSearch
# from langchain_google_community.tools import GoogleCustomSearch

class WebSearchService:
    def __init__(self):
        self.search_providers = {}
        self.search_history = []
        self.max_history_items = 50
        self.max_results_per_search = 10

        self._initialize_search_providers()

    def _initialize_search_providers(self):
        # Tavily Search (推奨 - AIに最適化された検索)
        if os.getenv("TAVILY_API_KEY"):
            try:
                from langchain_community.tools import TavilySearchResults
                self.search_providers["tavily"] = TavilySearchResults(
                    max_results=self.max_results_per_search,
                    api_key=os.getenv("TAVILY_API_KEY")
                )
            except ImportError:
                print("Warning: TavilySearchResults not installed. Please install langchain-community.")

        # Google Custom Search
        if os.getenv("GOOGLE_SEARCH_API_KEY") and os.getenv("GOOGLE_CSE_ID"):
            try:
                from langchain_community.tools.google_search import GoogleSearchAPIWrapper
                self.search_providers["google"] = GoogleSearchAPIWrapper(
                    google_api_key=os.getenv("GOOGLE_SEARCH_API_KEY"),
                    google_cse_id=os.getenv("GOOGLE_CSE_ID")
                )
            except ImportError:
                print("Warning: GoogleSearchAPIWrapper not installed. Please install langchain-community.")

        # DuckDuckGo (APIキー不要、フォールバック用)
        try:
            from langchain_community.tools import DuckDuckGoSearch
            self.search_providers["duckduckgo"] = DuckDuckGoSearch(
                max_results=self.max_results_per_search
            )
        except ImportError:
            print("Warning: DuckDuckGoSearch not installed. Please install langchain-community.")


        print(f"🔍 WebSearchService: Initialized {len(self.search_providers)} search providers")

    async def perform_search(self, query: str, options: dict = None):
        if options is None:
            options = {}

        provider_pref = options.get("provider", "auto")
        max_results = options.get("maxResults", self.max_results_per_search)
        filter_domains = options.get("filterDomains", [])
        exclude_domains = options.get("excludeDomains", [])
        language = options.get("language", "ja")
        region = options.get("region", "jp")

        try:
            print(f"🔍 WebSearchService: Performing search with query: \"{query}\"")

            selected_provider_instance = self._select_provider(provider_pref)
            if not selected_provider_instance:
                raise ValueError("利用可能な検索プロバイダーがありません")

            start_time = datetime.datetime.now()
            # LangChainのinvokeメソッドは同期的に動作する場合があるため、awaitを付ける
            raw_results = await selected_provider_instance.ainvoke(query)
            search_time = (datetime.datetime.now() - start_time).total_seconds() * 1000

            formatted_results = self._format_search_results(raw_results, {
                "maxResults": max_results,
                "filterDomains": filter_domains,
                "excludeDomains": exclude_domains
            })

            self._add_to_history({
                "query": query,
                "provider": self._get_provider_name(selected_provider_instance),
                "results": formatted_results,
                "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
                "searchTime": search_time
            })

            print(f"✅ WebSearchService: Found {len(formatted_results)} results in {search_time:.2f}ms")

            return {
                "success": True,
                "query": query,
                "results": formatted_results,
                "metadata": {
                    "provider": self._get_provider_name(selected_provider_instance),
                    "searchTime": search_time,
                    "totalResults": len(formatted_results),
                    "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat()
                }
            }

        except Exception as e:
            print(f"❌ WebSearchService: Search failed: {e}")
            return {
                "success": False,
                "query": query,
                "results": [],
                "error": str(e),
                "metadata": {
                    "provider": provider_pref,
                    "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat()
                }
            }

    def _select_provider(self, preference: str):
        if preference == "auto":
            if "tavily" in self.search_providers:
                return self.search_providers["tavily"]
            if "google" in self.search_providers:
                return self.search_providers["google"]
            return self.search_providers.get("duckduckgo")
        
        return self.search_providers.get(preference)

    def _get_provider_name(self, provider_instance):
        for name, instance in self.search_providers.items():
            if instance == provider_instance:
                return name
        return "unknown"

    def _format_search_results(self, raw_results, options: dict):
        max_results = options.get("maxResults", self.max_results_per_search)
        filter_domains = options.get("filterDomains", [])
        exclude_domains = options.get("excludeDomains", [])

        results = []
        if isinstance(raw_results, list):
            results = raw_results
        elif hasattr(raw_results, 'results') and isinstance(raw_results.results, list):
            results = raw_results.results
        elif isinstance(raw_results, str):
            # 文字列結果の場合（DuckDuckGo等）
            return [{
                "title": "Search Result",
                "url": "",
                "snippet": raw_results,
                "source": "text_result"
            }]

        formatted = []
        for result in results:
            url = result.get("url") or result.get("link")
            if not url:
                continue

            # ドメインフィルタリング
            if filter_domains and not any(domain in url for domain in filter_domains):
                continue
            if exclude_domains and any(domain in url for domain in exclude_domains):
                continue

            formatted.append({
                "title": result.get("title") or result.get("name") or "No Title",
                "url": url,
                "snippet": result.get("snippet") or result.get("description") or result.get("content") or "",
                "source": self._extract_domain(url),
                "score": result.get("score", 0),
                "publishedDate": result.get("published_date")
            })
            if len(formatted) >= max_results:
                break
        return formatted

    def _extract_domain(self, url: str):
        try:
            return urlparse(url).hostname
        except ValueError:
            return "unknown"

    def _add_to_history(self, search_entry: dict):
        self.search_history.append(search_entry)
        if len(self.search_history) > self.max_history_items:
            self.search_history = self.search_history[-self.max_history_items:]

    def get_search_history(self, limit: int = 10):
        return self.search_history[-limit:][::-1] # 新しい順

    def clear_search_history(self):
        self.search_history = []

    def get_available_providers(self):
        providers = {}
        for name in self.search_providers.keys():
            providers[name] = {
                "name": self._get_provider_display_name(name),
                "available": True,
                "requiresApiKey": self._requires_api_key(name)
            }
        return providers

    def _get_provider_display_name(self, name: str):
        display_names = {
            "tavily": "Tavily AI Search",
            "google": "Google Custom Search",
            "duckduckgo": "DuckDuckGo"
        }
        return display_names.get(name, name)

    def _requires_api_key(self, name: str):
        return name != "duckduckgo"

    def generate_search_summary(self, search_results: list, query: str):
        if not search_results:
            return f"\"{query}\" に関する検索結果は見つかりませんでした。"

        total_results = len(search_results)
        top_sources = []
        for result in search_results[:3]:
            source = result.get("source")
            if source and source not in top_sources:
                top_sources.append(source)

        return f"\"{query}\" について {total_results} 件の検索結果が見つかりました。主な情報源: {'. '.join(top_sources)}"

    def get_status(self):
        return {
            "availableProviders": list(self.search_providers.keys()),
            "searchHistoryCount": len(self.search_history),
            "maxResultsPerSearch": self.max_results_per_search,
            "isHealthy": len(self.search_providers) > 0,
            "lastActivity": self.search_history[-1]["timestamp"] if self.search_history else None
        }

File: server/tool/web-search-service.js
/* =========================================
    Web検索サービス（LangChain統合）
   ========================================= */

/*
## 概要
LangChainを使用したWeb検索機能を提供する専門サービス。
複数の検索プロバイダーを統合し、結果の整形・フィルタリングを行う。

## 責任
- LangChainを使った検索実行
- 検索結果の整形・要約
- 検索ソースの管理
- 検索履歴の保持
*/

import { TavilySearchResults } from "@langchain/community/tools/tavily_search";
import { DuckDuckGoSearch } from "@langchain/community/tools/duckduckgo_search";
import { GoogleCustomSearch } from "@langchain/community/tools/google_custom_search";

export class WebSearchService {
    constructor() {
        this.searchProviders = new Map();
        this.searchHistory = [];
        this.maxHistoryItems = 50;
        this.maxResultsPerSearch = 10;
        
        // 利用可能な検索プロバイダーを初期化
        this._initializeSearchProviders();
    }

    /**
     * 検索プロバイダーの初期化
     */
    _initializeSearchProviders() {
        // Tavily Search (推奨 - AIに最適化された検索)
        if (process.env.TAVILY_API_KEY) {
            this.searchProviders.set('tavily', new TavilySearchResults({
                maxResults: this.maxResultsPerSearch,
                apiKey: process.env.TAVILY_API_KEY
            }));
        }

        // Google Custom Search
        if (process.env.GOOGLE_SEARCH_API_KEY && process.env.GOOGLE_CSE_ID) {
            this.searchProviders.set('google', new GoogleCustomSearch({
                apiKey: process.env.GOOGLE_SEARCH_API_KEY,
                googleCSEId: process.env.GOOGLE_CSE_ID
            }));
        }

        // DuckDuckGo (APIキー不要、フォールバック用)
        this.searchProviders.set('duckduckgo', new DuckDuckGoSearch({
            maxResults: this.maxResultsPerSearch
        }));

        console.log(`🔍 WebSearchService: Initialized ${this.searchProviders.size} search providers`);
    }

    /**
     * Web検索を実行
     */
    async performSearch(query, options = {}) {
        const {
            provider = 'auto',
            maxResults = this.maxResultsPerSearch,
            filterDomains = [],
            excludeDomains = [],
            language = 'ja',
            region = 'jp'
        } = options;

        try {
            console.log(`🔍 WebSearchService: Performing search with query: "${query}"`);

            // 検索プロバイダーを選択
            const selectedProvider = this._selectProvider(provider);
            if (!selectedProvider) {
                throw new Error('利用可能な検索プロバイダーがありません');
            }

            // 検索実行
            const startTime = Date.now();
            const rawResults = await selectedProvider.invoke(query);
            const searchTime = Date.now() - startTime;

            // 結果を整形
            const formattedResults = this._formatSearchResults(rawResults, {
                maxResults,
                filterDomains,
                excludeDomains
            });

            // 検索履歴に保存
            this._addToHistory({
                query,
                provider: this._getProviderName(selectedProvider),
                results: formattedResults,
                timestamp: new Date().toISOString(),
                searchTime
            });

            console.log(`✅ WebSearchService: Found ${formattedResults.length} results in ${searchTime}ms`);

            return {
                success: true,
                query,
                results: formattedResults,
                metadata: {
                    provider: this._getProviderName(selectedProvider),
                    searchTime,
                    totalResults: formattedResults.length,
                    timestamp: new Date().toISOString()
                }
            };

        } catch (error) {
            console.error('❌ WebSearchService: Search failed:', error);
            return {
                success: false,
                query,
                results: [],
                error: error.message,
                metadata: {
                    provider: provider,
                    timestamp: new Date().toISOString()
                }
            };
        }
    }

    /**
     * 検索プロバイダーの選択
     */
    _selectProvider(preference) {
        if (preference === 'auto') {
            // 優先順位: Tavily > Google > DuckDuckGo
            if (this.searchProviders.has('tavily')) {
                return this.searchProviders.get('tavily');
            }
            if (this.searchProviders.has('google')) {
                return this.searchProviders.get('google');
            }
            return this.searchProviders.get('duckduckgo');
        }

        return this.searchProviders.get(preference) || this.searchProviders.get('duckduckgo');
    }

    /**
     * プロバイダー名を取得
     */
    _getProviderName(provider) {
        for (const [name, providerInstance] of this.searchProviders.entries()) {
            if (providerInstance === provider) {
                return name;
            }
        }
        return 'unknown';
    }

    /**
     * 検索結果の整形
     */
    _formatSearchResults(rawResults, options = {}) {
        const { maxResults, filterDomains, excludeDomains } = options;
        
        let results = [];
        
        // LangChainの結果形式に応じて処理
        if (Array.isArray(rawResults)) {
            results = rawResults;
        } else if (rawResults.results && Array.isArray(rawResults.results)) {
            results = rawResults.results;
        } else if (typeof rawResults === 'string') {
            // 文字列結果の場合（DuckDuckGo等）
            return [{
                title: 'Search Result',
                url: '',
                snippet: rawResults,
                source: 'text_result'
            }];
        }

        return results
            .filter(result => result && result.url)
            .filter(result => {
                // ドメインフィルタリング
                if (filterDomains.length > 0) {
                    return filterDomains.some(domain => result.url.includes(domain));
                }
                if (excludeDomains.length > 0) {
                    return !excludeDomains.some(domain => result.url.includes(domain));
                }
                return true;
            })
            .slice(0, maxResults)
            .map(result => ({
                title: result.title || result.name || 'No Title',
                url: result.url || result.link || '',
                snippet: result.snippet || result.description || result.content || '',
                source: this._extractDomain(result.url || result.link || ''),
                score: result.score || 0,
                publishedDate: result.published_date || null
            }));
    }

    /**
     * URLからドメイン名を抽出
     */
    _extractDomain(url) {
        try {
            return new URL(url).hostname;
        } catch (error) {
            return 'unknown';
        }
    }

    /**
     * 検索履歴に追加
     */
    _addToHistory(searchEntry) {
        this.searchHistory.push(searchEntry);
        
        // 履歴数制限
        if (this.searchHistory.length > this.maxHistoryItems) {
            this.searchHistory = this.searchHistory.slice(-this.maxHistoryItems);
        }
    }

    /**
     * 検索履歴を取得
     */
    getSearchHistory(limit = 10) {
        return this.searchHistory
            .slice(-limit)
            .reverse(); // 新しい順
    }

    /**
     * 検索履歴をクリア
     */
    clearSearchHistory() {
        this.searchHistory = [];
    }

    /**
     * 利用可能な検索プロバイダー一覧
     */
    getAvailableProviders() {
        const providers = {};
        
        for (const [name, provider] of this.searchProviders.entries()) {
            providers[name] = {
                name: this._getProviderDisplayName(name),
                available: true,
                requiresApiKey: this._requiresApiKey(name)
            };
        }

        return providers;
    }

    /**
     * プロバイダー表示名を取得
     */
    _getProviderDisplayName(name) {
        const displayNames = {
            'tavily': 'Tavily AI Search',
            'google': 'Google Custom Search',
            'duckduckgo': 'DuckDuckGo'
        };
        return displayNames[name] || name;
    }

    /**
     * APIキーが必要かどうかを判定
     */
    _requiresApiKey(name) {
        return name !== 'duckduckgo';
    }

    /**
     * 検索結果のサマリーを生成
     */
    generateSearchSummary(searchResults, query) {
        if (!searchResults || searchResults.length === 0) {
            return `"${query}" に関する検索結果は見つかりませんでした。`;
        }

        const totalResults = searchResults.length;
        const topSources = searchResults
            .slice(0, 3)
            .map(result => result.source)
            .filter((source, index, self) => self.indexOf(source) === index);

        return `"${query}" について ${totalResults} 件の検索結果が見つかりました。主な情報源: ${topSources.join(', ')}`;
    }

    /**
     * サービスの状態を取得
     */
    getStatus() {
        return {
            availableProviders: Object.keys(Object.fromEntries(this.searchProviders)),
            searchHistoryCount: this.searchHistory.length,
            maxResultsPerSearch: this.maxResultsPerSearch,
            isHealthy: this.searchProviders.size > 0,
            lastActivity: this.searchHistory.length > 0 
                ? this.searchHistory[this.searchHistory.length - 1].timestamp 
                : null
        };
    }
}